Content.Server/_Impstation/Homunculi/HomunculusSystem.cs

```diff
@@ -0,0 +1,171 @@
+using Content.Server.Humanoid;
+using Content.Server._Impstation.Homunculi.Incubator;
+using Content.Shared.Chemistry.Components;
+using Content.Shared.Chemistry.EntitySystems;
+using Content.Shared.Chemistry.Reagent;
+using Content.Shared.Forensics.Components;
+using Content.Shared.Humanoid.Markings;
+using Content.Shared.Humanoid;
+using Content.Shared._Impstation.Homunculi.Components;
+using Content.Shared._Impstation.Homunculi.Incubator.Components;
+using Robust.Server.GameObjects;
+using Robust.Shared.Map;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+
+namespace Content.Server._Impstation.Homunculi;
+
+public sealed class HomunculusSystem : EntitySystem
+{
+    [Dependency] private readonly HumanoidAppearanceSystem _appearance = default!;
+    [Dependency] private readonly SharedSolutionContainerSystem _solution = default!;
+    [Dependency] private readonly TransformSystem _transform = default!;
+    [Dependency] private readonly IncubatorSystem _incubator = default!;
+
+    public bool CreateHomunculiWithDna(Entity<IncubatorComponent?> ent, Entity<SolutionComponent> solution, MapCoordinates mapCoordinates, [NotNullWhen(true)] out EntityUid? homunculus)
+    {
+        // If there's no DNA data in the solution, return
+        if (!_incubator.HasDnaData(solution))
+        {
+            homunculus = null;
+            return false;
+        }
+
+        // Save a copy of the solutions reagents, can't just use it straight up
+        var reagentList = solution.Comp.Solution.Contents.ToList();
+        List<string?> storedDna = [];
+        List<Entity<HomunculiTypeComponent>> entities = [];
+
+        foreach (var dnaList in reagentList.Select(reagent => reagent.Reagent.EnsureReagentData().OfType<DnaData>()))
+        {
+            storedDna.AddRange(dnaList.Select(dna => dna.DNA));
+        }
+
+        var query = EntityQueryEnumerator<HomunculiTypeComponent>();
+        while (query.MoveNext(out var entityUid,out var homunculiTypeComponent))
+        {
+            if (!TryComp<DnaComponent>(entityUid, out var dna))
+                continue;
```

Comment:
```suggestion
        var query = EntityQueryEnumerator<HomunculiTypeComponent, DnaComponent>();
        while (query.MoveNext(out var entityUid ,out var homunculiTypeComponent, out var DnaComponent))
        {
            
```
can just use a multi-comp query enumerator here.
also generally the vibes on this bit of code are quite bad but for what it's doing I think this is the only way to do it?

Reply:
it is in fact the only way to do it sadly